integrate.js

```diff
@@ -0,0 +1,164 @@
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+"use strict";
+
+(function(Nuvola)
+{
+	// Create media player component
+	var player = Nuvola.$object(Nuvola.MediaPlayer);
+
+	// Handy aliases
+	var PlaybackState = Nuvola.PlaybackState;
+	var PlayerAction = Nuvola.PlayerAction;
+
+	// Create new WebApp prototype
+	var WebApp = Nuvola.$WebApp();
+
+	// Initialization routines
+	WebApp._onInitWebWorker = function(emitter)
+	{
+		Nuvola.WebApp._onInitWebWorker.call(this, emitter);
+
+		var state = document.readyState;
+		if (state === "interactive" || state === "complete")
+			this._onPageReady();
+		else
+			document.addEventListener("DOMContentLoaded", this._onPageReady.bind(this));
+	}
+
+	WebApp._onPageReady = function()
+	{
+		// Connect handler for signal ActionActivated
+		Nuvola.actions.connect("ActionActivated", this);
+
+		// Start update routine
+		this.update();
+	}
+
+	// Extract data from the web page
+	WebApp.update = function()
+	{
+		var can_prev = false;
+		var can_next = false;
+		var can_play = false;
+		var can_pause = false;
+		var track = {
+			title:       null,
+			artist:      null,
+			album:       null,
+			artLocation: null
+		}
+
+		try
+		{
+			var playingTrack = R.Services.Player.model.get("playingTrack").attributes;
+			track = {
+				title:  playingTrack.name,
+				artist: playingTrack.artist,
+				album: playingTrack.album,
+				artLocation: playingTrack.baseIcon
+			}
+		}
+		catch (e)
+		{
+			console.debug(Nuvola.format(Nuvola.Translate.gettext("Unable to obtain song info: {1}"), e.message));
+		}
+
+		var state = PlaybackState.UNKNOWN;
+		try
+		{
+			switch (R.Services.Player.model.attributes.playState)
+			{
+				case 0:
+					state = PlaybackState.PAUSED;
+					can_pause = false;
+					can_play = true;
+					break;
+				case 1:
+					state = PlaybackState.PLAYING;
+					can_pause = true;
+					can_play = false;
+					break;
+			}
+
+			if (state != PlaybackState.UNKNOWN)
+				can_prev = can_next = true;
+		}
+		catch (e)
+		{
+			// console.debug("Unable to obtain state info: " + e.message);
+			can_prev = can_next = false;
+		}
+
+		player.setTrack(track);
+		player.setPlaybackState(state);
+		player.setCanGoNext(can_next);
+		player.setCanGoPrev(can_prev);
+		player.setCanPlay(can_play);
+		player.setCanPause(can_pause);
+
+		// Schedule the next update
+		setTimeout(this.update.bind(this), 500);
+	}
+
+	// Handler of playback actions
+	WebApp._onActionActivated = function(emitter, name, param)
+	{
```

Comment:
When Rdio starts for the first time, variable `R` is not defined. However, GNOME Shell media player extension (not sure about Unity sound indicator) shows a play button even though `player.setCanPlay()` was properly set to `false`. If the play button is clicked, you integration produces error: "Error during call of Nuvola.actions.emit: Function 'Nuvola.actions.emit' failed. : Can't find variable: R". Therefore, you have to check whether `R` is available before using it.